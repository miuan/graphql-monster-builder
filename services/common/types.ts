export type SchemaModel = {
    modelName: string
    start: number
    end: number
    members: SchemaModelMember[]
    protection: SchemaModelProtection
    type: SchemaModelType
}

export enum SchemaModelType {
    MODEL = 'MODEL', // me as one to many
    ENTITY = 'ENTITY', // me as multi to one
}

export enum SchemaModelRelationType {
    RELATION = 'RELATION',
    CONNECT = 'CONNECT',
    ENTITY = 'ENTITY',
}

export type SchemaModelRelation = {
    name: string
    type: SchemaModelRelationType
    inputName: string
    relatedModel: SchemaModel
    relatedMember: SchemaModelMember
    createFromAnotherModel: boolean
    payloadNameForCreate: string
    payloadNameForId: string
    payloadNameForAddOrDelete: string
    error: string
}

export type SchemaModelMember = {
    name: string
    type: string
    modelName: string
    isRequired: boolean
    isUnique: boolean | string[]
    relation: SchemaModelRelation
    isArray: boolean
    row: number
    isReadonly: boolean
    default?: number | string
    regExp?: string
    placeholder?: string
}

export enum SchemaModelProtectionType {
    PUBLIC,
    USER,
    OWNER,
    ROLE,
    FILTER,
}

export type SchemaModelProtectionParam = {
    type: SchemaModelProtectionType
    typeName?: string
    roles?: string[]
    filter?: SchemaModelProtectionFilter[]
    whitelist?: string[]
    blacklist?: string[]
    param?: string
}

export type SchemaModelProtectionFilter = {
    name: string
    value: string
}

export type SchemaModelProtection = {
    all: SchemaModelProtectionParam[]
    one: SchemaModelProtectionParam[]
    create: SchemaModelProtectionParam[]
    update: SchemaModelProtectionParam[]
    remove: SchemaModelProtectionParam[]
}

export type StructureItem = {
    dir: string
    modules: string[]
}

export type Structure = {
    id: string
}

export type StructureBackend = Structure & {
    index: StructureItem
    schema: StructureItem
    gen: StructureItem
    models: StructureItem
    resolvers: StructureItem
    services: StructureItem
}

export type StructureFrontend = Structure & {
    index: StructureItem
    gen: StructureItem
    graphql: StructureItem
}

export const MODEL_TYPES = ['type', 'model', 'entity']
export const MODELS_NOT_HAVE_CREATE = ['User', 'File']
