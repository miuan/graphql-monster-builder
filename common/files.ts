const fs = require('fs');
const path = require('path');
import {
  Structure,
  StructureItem,
} from '../common/types';

export const STRUCTURE = {
  schema: { 
    dir: `ID/`,
    modules: ['schema.protectql'],
  },
  gen: { 
    dir: `ID/gen/`,
    modules: [],
  },
  models: { 
    dir:`ID/gen/models/`, 
    modules: [],
  },
  resolvers: { 
    dir:`ID/gen/resolvers/`, 
    modules: [],
  },
  services: { 
    dir:'ID/gen/services/', 
    modules: [],
  },
};

export const removeDirs = async (dir) => {
  if(!fs.existsSync(dir)){
    return
  } else if(!fs.lstatSync(dir).isDirectory()) {
    return fs.unlinkSync(dir);
  }
  
  if(/node_modules$/.test(dir)) {
    console.log(`skip dir: ${dir}`)
    return
  }

  console.log(`remove dir: '${dir}'`);
  const files = fs.readdirSync(dir);
  for (const file of files) {
    const fullfile = path.join(dir, file);
    const stats = fs.statSync(fullfile);
    if (stats.isDirectory()) {
      removeDirs(fullfile);
    } else if(! /yarn.lock$/.test(fullfile) && ! /package.json$/.test(fullfile)) {
      // also not remove 'yarn.lock
      fs.unlinkSync(fullfile);
    } else {
      console.log('skiped: ', fullfile)
    }
  }

  // if node modules stays then rm on root dir will fail
  try {
    fs.rmdirSync(dir);
  } catch (ex) {

  }
  
};

export const createDirs = (structure: Structure, baseDir: string = '') => {
  
  for (const str in structure) {
    if (str === 'id') {
      continue;
    }

    const dir = path.join(baseDir, structure[str].dir);
    // if node modules stays then rm on root dir was not removed
    try {
      fs.mkdirSync(dir);
    } catch (ex) {

    }
    
  }
};

export const generateStructure = (id, baseDir: string = './graphql/generated/'): Structure => {
  const structure: Structure = JSON.parse(JSON.stringify(STRUCTURE)) as Structure;
  
  structure.id = id;

  for (const str in structure) {
    if (str === 'id') {
      continue;
    }

    const item = structure[str];
    item.dir = path.join(baseDir, item.dir.replace('ID', id));
  }

  removeDirs(path.join(baseDir, id));
  createDirs(structure);

  return structure;
};


export const writeToFile = (item: StructureItem, name: string, content: string) => {
  const nameWithExt = name.indexOf('.') === -1 ? `${name}.ts` : name;
  const fullName = path.join(item.dir, nameWithExt);

  // some files should not be overrided 
  // for example the package.json 
  if(!fs.existsSync(fullName)){
    fs.writeFileSync(fullName, content);
    item.modules.push(name);
  }
};


export const templateToText = (templateName: string, params: any = null): string => {
  const templateFilePath = `./bd/templates/${templateName}`;
  let text = fs.readFileSync(templateFilePath).toString();
  
  if (params && params instanceof Object) {
    for (const key in params) {
      const search = new RegExp(key, 'g');
      text = text.replace(search, params[key]);
    }
  }

  return text;
};

export const templateToFile = (templateName: string, params: any) => {

};
