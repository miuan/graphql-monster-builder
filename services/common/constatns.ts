export const schemaFilterNumberValue = [
    ['not', 'Values that are not equal to given value.'],
    ['in', 'All values that are contained in given list.', true],
    ['not_in', 'All values that are not contained in given list.', true],
    ['lt', 'All values less than the given value.'],
    ['lte', 'All values less than or equal the given value.'],
    ['gt', 'All values greater than the given value.'],
    ['gte', 'All values greater than or equal the given value.'],
]

export const schemaFilterStringValue = [
    ...schemaFilterNumberValue,
    ['contains'],
    ['not_contains'],
    ['starts_with'],
    ['not_starts_with'],
    ['ends_with'],
    ['not_ends_with'],
]

export const schemaFilterArrays = ['every', 'some', 'none']

export const schemaMutations = [
    ['create', 'NO_ID'],
    ['update', 'ALWAYS_ID'],
    // 'updateOrCreate',
    ['remove', 'ONLY_ID'],
]

export const schemaMutationsForRelations = [
    ['addTo', 'RELATION_MEMBER_ONE_TO_MANY'],
    ['removeFrom', 'RELATION_MEMBER_ONE_TO_MANY'],
]

export const schemaOrderBy = ['ASC', 'DESC']

export const notMutationFields = {
    '': ['id', 'createdAt', 'updatedAt'],
    File: ['contentType', 'secret', 'size', 'url'],
}

export const MONGOSEE_RESERVED_WORDS = [
    'on',
    'emit',
    '_events',
    'db',
    'get',
    'set',
    'init',
    'isNew',
    'errors',
    'schema',
    'options',
    'modelName',
    'collection',
    '_pres',
    '_posts',
    'toObject',
]
