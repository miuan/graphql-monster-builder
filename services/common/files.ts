const fs = require('fs');
const path = require('path');
import loglevel from '../log'
import { setupModelsRelations } from '../parser/relations';
const log = loglevel.getLogger('files')

import {
  Structure,
  StructureItem,
} from './types';



export const removeDirs = (dir, force = false) => {
  if(!fs.existsSync(dir)){
    return
  } else if(!fs.lstatSync(dir).isDirectory()) {
    return fs.unlinkSync(dir);
  }
  
  if(/node_modules$/.test(dir)) {
    log.trace(`skip dir: ${dir}`)
    return
  }

  if(/^.git/.test(dir)) {
    log.trace(`skip dir: ${dir}`)
    return
  }


  log.info(`remove dir: '${dir}'`);
  const files = fs.readdirSync(dir);
  for (const file of files) {
    const fullfile = path.join(dir, file);
    const stats = fs.statSync(fullfile);
    if (stats.isDirectory()) {
      removeDirs(fullfile);
    } else if( ! /yarn.lock$/.test(fullfile) 
               ) {
      // also not remove 'yarn.lock
      fs.unlinkSync(fullfile);
    } else {
      log.trace('skiped: ', fullfile)
    }
  }

  // if node modules stays then rm on root dir will fail
  try {
    fs.rmdirSync(dir);
  } catch (ex) {

  }
  
};

export const createDirs = (structure: Structure, baseDir: string = '') => {
  
  for (const str in structure) {
    if (str === 'id') {
      continue;
    }

    const dir = path.join(baseDir, structure[str].dir);
    // if node modules stays then rm on root dir was not removed
    try {
      fs.mkdirSync(dir, { recursive: true })
      log.info(`Create dir: ${dir}`)
    } catch (ex) {

    }
    
  }
};


export const writeToFile = (item: StructureItem, name: string, content: string): boolean => {
  const nameWithExt = name.indexOf('.') === -1 ? `${name}.ts` : name;
  const fullName = path.join(item.dir, nameWithExt);

  // some files should not be overrided 
  // for example the package.json 
  if(!fs.existsSync(fullName)){
    fs.writeFileSync(fullName, content)
    item.modules.push(name);
  }

  return true
};



export const templateToText = (text: string, params: any) => {
  if (params && params instanceof Object) {
    for (const key in params) {
      const search = new RegExp(key, 'g');
      text = text.replace(search, params[key]);
    }
  }

  return text
}

export const templateFileToText = (fileName: string, params: any = null): string => {
  const templateFilePath = path.join(__dirname, '../../', 'services/backend/templates', fileName)
  //const templateFilePath = `services/backend/templates/${fileName}`;
  let text = fs.readFileSync(templateFilePath).toString();
  
  return templateToText(text, params);
};

export const templateToFile = (templateName: string, params: any) => {

};
