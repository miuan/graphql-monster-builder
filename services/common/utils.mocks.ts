import { testResolverUpdate } from "../backend/templates/__generated_resolvers_for_test__/testResolver_filter"
import { SchemaModelMember, SchemaModel, SchemaModelRelationType } from "./types"

export const modelUserWithTodoManyToManyRelation = {
    modelName: 'User',
    members: [{
        modelName: 'Todo',
        name: 'Todo',
        type: 'Todo',
        isArray: true,
        relation: {
            type: SchemaModelRelationType.RELATION,
            name: 'UserOnTodo',
        } 
    }] as SchemaModelMember[]
} as SchemaModel

export const modelTodoWithUserManyToManyRelation = {
    modelName: 'Todo',
    members: [{
        modelName: 'User',
        name: 'User',
        type: 'User',
        isArray: true,
        relation: {
            type: SchemaModelRelationType.RELATION,
            name: 'UserOnTodo',
            relatedModel: modelUserWithTodoManyToManyRelation,
        } 
    }] as SchemaModelMember[]
} as SchemaModel

modelUserWithTodoManyToManyRelation.members[0].relation.relatedModel = modelTodoWithUserManyToManyRelation
modelUserWithTodoManyToManyRelation.members[0].relation.relatedMember = modelTodoWithUserManyToManyRelation.members[0]
modelTodoWithUserManyToManyRelation.members[0].relation.relatedModel = modelUserWithTodoManyToManyRelation
modelTodoWithUserManyToManyRelation.members[0].relation.relatedMember = modelUserWithTodoManyToManyRelation.members[0]

export const modelUserWithTodoOneToManyRelation = {
    modelName: 'User',
    members: [{
        name: 'Todo',
        type: 'Todo',
        isArray: false,
        relation: {
            type: SchemaModelRelationType.RELATION,
            name: 'UserOnTodo',
        } 
    }] as SchemaModelMember[]
} as SchemaModel

export const modelTodoWithUserManyToOneRelation = {
    modelName: 'User',
    members: [{
        name: 'Todo',
        type: 'Todo',
        isArray: true,
        relation: {
            type: SchemaModelRelationType.RELATION,
            name: 'UserOnTodo',
        } 
    }] as SchemaModelMember[]
} as SchemaModel

modelUserWithTodoOneToManyRelation.members[0].relation.relatedModel = modelTodoWithUserManyToOneRelation
modelUserWithTodoOneToManyRelation.members[0].relation.relatedMember = modelTodoWithUserManyToOneRelation.members[0]
modelTodoWithUserManyToOneRelation.members[0].relation.relatedModel = modelUserWithTodoOneToManyRelation
modelTodoWithUserManyToOneRelation.members[0].relation.relatedMember = modelUserWithTodoOneToManyRelation.members[0]

export const relationUserOnTodo = {
    name: 'Todo',
    type: 'Todo',
    modelName: 'Todo',
    isArray: false,
    relation: {
        type: SchemaModelRelationType.RELATION,
        relatedModel: modelUserWithTodoManyToManyRelation,
        name: 'UserOnTodo',
    } 
} as SchemaModelMember

export const relationUserOnPress = {
    relation: {
        name: 'UserOnPress',
    } 
} as SchemaModelMember

export const relationUserOnWall = {
    relation: {
        name: 'UserOnWall',
    } 
} as SchemaModelMember

