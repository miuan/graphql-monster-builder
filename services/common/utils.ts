import { SchemaModel, SchemaModelRelationType, SchemaModelMember } from './types'
import logger from '../log'
const log = logger.getLogger('utils')

export const searchModelsRelationsInModels = (relationName: string, models: SchemaModel[], exclude: string[] = []) => {
    for (const model of models) {
        if (exclude.indexOf(model.modelName) === -1) {
            for (const member of model.members) {
                if (member.relation && member.relation.name === relationName) {
                    return { model, member }
                }
            }
        }
    }

    return null
}

// in many to many relation we need create only one add or delete methods
// for example User and UserRole we want only userAddToUserRole and userRemoveFromUserRole
// but because the relation is cross each others it will create also userRoleAddToUser and userRoleRemoveFromUser
export const getOnlyOneRelatedMember = (member: SchemaModelMember) => {
    const {
        relation: { name: relationName, relatedMember },
    } = member

    if (member.isArray && (!relatedMember.isArray || member.modelName > relatedMember?.modelName)) {
        return member
    }

    // if any of them is not array, not take any
    return null
}

export const firstToLower = (name) => name.charAt(0).toLowerCase() + name.slice(1)
export const firstToUpper = (name) => name.charAt(0).toUpperCase() + name.slice(1)
