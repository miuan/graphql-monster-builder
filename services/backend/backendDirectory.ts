import path = require("path")
import * as files from "../common/files"
import { Structure } from "../common/types"
  
export const BACKEND_STRUCTURE = {
    schema: { 
        dir: `ID/`,
        modules: ['schema.protectql'],
    },
    gen: { 
        dir: `ID/gen/`,
        modules: [],
    },
    models: { 
        dir:`ID/gen/models/`, 
        modules: [],
    },
    resolvers: { 
        dir:`ID/gen/resolvers/`, 
        modules: [],
    },
    services: { 
        dir:'ID/gen/services/', 
        modules: [],
    },
};
  
export interface BackendStructureOperatorWrite {
    write: (file: string, data: string) => boolean
    genWrite: (file: string, data: string) => boolean
    modelsWrite: (file: string, data: string) => boolean
    resolversWrite: (file: string, data: string) => boolean
    servicesWrite: (file: string, data: string) => boolean
}

export class BackendDirectory implements BackendStructureOperatorWrite {
    structure: Structure
    baseDir: string
    projectName: string

    constructor() {
        this.structure = JSON.parse(JSON.stringify(BACKEND_STRUCTURE)) as Structure
    }

    /**
     * 
     * @param projectName project name
     * @param baseDir base dir
     */
    init(projectName: string, baseDir: string = './graphql/generated/'){
        this.projectName = projectName
        this.baseDir = baseDir

        const outDir = projectName
        this.structure.id = projectName.replace('../', '');
      
        for (const str in this.structure) {
          if (str === 'id') {
            continue;
          }
      
          const item = this.structure[str];
          item.dir = path.join(baseDir, item.dir.replace('ID', outDir));
        }
    }

    /**
     * create or clean directory for copy files
     */
    prepareDirectory(){
        if(!this.baseDir || !this.projectName) {
            throw new Error('this.projectName or this.baseDir are empty, call init() fist')
        }

        files.removeDirs(path.join(this.baseDir, this.projectName, 'gen'));
        files.createDirs(this.structure)
    }
    
    public write(file: string, data: string){
        return files.writeToFile(this.structure.schema, file, data)
    }

    public genWrite(file: string, data: string){
        return files.writeToFile(this.structure.gen, file, data)
    }

    modelsWrite(file: string, data: string){
        return files.writeToFile(this.structure.models, file, data)
    }

    resolversWrite(file: string, data: string){
        return files.writeToFile(this.structure.resolvers, file, data)
    }

    servicesWrite(file: string, data: string){
        return files.writeToFile(this.structure.services, file, data)
    }
  
  }