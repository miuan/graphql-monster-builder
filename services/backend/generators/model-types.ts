import {
  SchemaModel,
  SchemaModelMember,
} from '../../common/types';


const defaultMembers = [
  'createdAt',
  'updatedAt',
  'id',
];

import logger from '../../log'
import { BackendDirectory } from '../backendDirectory';
const log = logger.getLogger('types')

export const createModelTypeFromModel = (model : SchemaModel) => {
  const {modelName} = model
  let modelMembers = ''

  for (const member of model.members) {
    modelMembers += `\n\t\t${member.name}: ${transformMemberTypeToTypescriptType(member)}`
  }

  if (modelName === 'User') {
    modelMembers += `\n\t\t__token?: string,\n`
    modelMembers += `\n\t\t__refreshToken?: string,\n`
    modelMembers += `\n\t\t__verifyToken?: string,\n`
    modelMembers += `\n\t\t__password: string,\n`
    modelMembers += `\n\t\t__resetPasswordToken: string,\n`
    modelMembers += `\n\t\t__parent_access_token: string,\n`
  }

  return `
    // type for model: \`${modelName}\`
    export interface ${modelName}Model extends Document {${modelMembers}
    }
  `
}
  
export const transformMemberTypeToTypescriptType = (member : SchemaModelMember) => {
  let ttype

  if (member.relation) {
     ttype = `${member.relation.relatedModel.modelName}Model`;
  } else if (member.type === 'DateTime') {
    ttype = 'Date';
  } else if (member.type === 'Int' || member.type === 'Float') {
    ttype =`number`;
  } else if (member.type === 'ID' || member.type === 'String') {
    ttype =`string`;
  } else {
    ttype = `${member.type}`;
  }

  if(member.isArray){
    ttype+='[]'
  }

  return ttype
};

export const generateModelTypes = (backendDirectory: BackendDirectory, models: SchemaModel[]) => {
  log.trace('generateTypescriptTypes')
  let typesAll = `import { Document } from 'mongoose'\n`
  for (const model of models) {
    typesAll += createModelTypeFromModel(model);
  }
  log.info(`Generate types for all models: types.ts`)
  backendDirectory.genWrite(`model-types.ts`, typesAll);
};
