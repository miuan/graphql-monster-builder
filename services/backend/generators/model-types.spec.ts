import { SchemaModel, SchemaModelMember, SchemaModelRelation, Structure, StructureItem } from "../../common/types"
import { createModelTypeFromModel, generateModelTypes, transformMemberTypeToTypescriptType } from "./model-types"
import {
    writeToFile,
  } from '../../common/files';
import { BackendDirectory } from "../backendDirectory";

describe('typescript-type', ()=>{
    describe('transformTypeToTypescriptType', ()=>{
        it.each([
            ['String', 'String'],
            ['DateTime', 'Date'],
            ['Int', 'Number'],
            ['Boolean', 'Boolean'],
            ['ID', 'String'],
        ])('should transform simple type %s to %s', (i, o)=>{
            expect(transformMemberTypeToTypescriptType({type: i} as SchemaModelMember)).toEqual(o)
        })

        it.each([
            ['String', 'String[]'],
            ['DateTime', 'Date[]'],
            ['Int', 'Number[]'],
            ['Boolean', 'Boolean[]'],
        ])('should transform simple array type %s to %s', (i, o)=>{
            expect(transformMemberTypeToTypescriptType({type: i, isArray: true} as SchemaModelMember)).toEqual(o)
        })

        it.each([
            ['Todo', 'Todo'],
            ['User', 'User'],
        ])('should transform related member %s to %s', (i, o)=>{
            expect(transformMemberTypeToTypescriptType({
                relation:{
                    relatedModel:{modelName: i}
                } as SchemaModelRelation
            } as SchemaModelMember)).toEqual(o)
        })

        it.each([
            ['Todo', 'Todo[]'],
            ['User', 'User[]'],
        ])('should transform related member array type %s to %s', (i, o)=>{
            expect(transformMemberTypeToTypescriptType({
                isArray: true,
                relation:{
                    relatedModel:{modelName: i}
                } as SchemaModelRelation
            } as SchemaModelMember)).toEqual(o)
        })
    })

    describe('createTypescriptType', ()=>{
        expect(createModelTypeFromModel({
            modelName: 'Todo',
            members: [{
                name: 'name',
                type: 'String'
            },{
                name: 'isDone',
                type: 'Boolean'
            }]
        } as SchemaModel)).toMatchSnapshot()
    })

    describe('generateTypescriptTypes', ()=>{
       
        it('should call 2x createTypescriptType and 1x writeToFile inside generateTypescriptTypes', ()=>{
            const backendDirectory = new BackendDirectory()
            backendDirectory.genWrite = jest.fn()

            generateModelTypes(backendDirectory, [{
                modelName: 'Todo',
                members: [{
                    name: 'name',
                    type: 'String'
                },{
                    name: 'isDone',
                    type: 'Boolean'
                }]
            } as SchemaModel,
            {
                modelName: 'User',
                members: [{
                    name: 'email',
                    type: 'String'
                },{
                    name: 'password',
                    type: 'Boolean'
                },{
                    name: 'todos',
                    isArray: true,
                    relation:{
                        relatedModel:{modelName: 'Todo'}
                    } as SchemaModelRelation
                }]
            } as SchemaModel])

            expect(backendDirectory.genWrite).toHaveBeenCalledTimes(1)
            expect((<any>backendDirectory.genWrite).mock.calls[0]).toMatchSnapshot()
        })

    })
})