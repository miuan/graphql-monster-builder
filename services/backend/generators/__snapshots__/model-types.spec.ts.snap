// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"
    // type for model: \`Todo\`
    export interface TodoModel extends Document {
		name: string
		isDone: Boolean
    }
  "
`;

exports[`typescript-type generateTypescriptTypes should call 2x createTypescriptType and 1x writeToFile inside generateTypescriptTypes 1`] = `
Array [
  "model-types.ts",
  "import { Document } from 'mongoose'

    // type for model: \`Todo\`
    export interface TodoModel extends Document {
		name: string
		isDone: Boolean
    }
  
    // type for model: \`User\`
    export interface UserModel extends Document {
		email: string
		password: Boolean
		todos: TodoModel[]
    }
  ",
]
`;
